src/Tools/IdeBenchmarks/SQLitePersistentStorageBenchmark.cs

```diff
@@ -0,0 +1,143 @@
+using System.Linq;
+using System.Threading.Tasks;
+using BenchmarkDotNet.Attributes;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.SolutionSize;
+using Microsoft.CodeAnalysis.SQLite;
+using Microsoft.CodeAnalysis.Storage;
+using Microsoft.CodeAnalysis.Test.Utilities;
+
+namespace IdeBenchmarks
+{
+    public class SQLitePersistentStorageBenchmarks
+    {
+        private readonly UseExportProviderAttribute _useExportProviderAttribute = new UseExportProviderAttribute();
+
+        // Run the test with different ratios of reads/writes.
+        [Params(0, 25, 50, 75, 100)]
+        public int ReadPercentage { get; set; }
+
+        private TestWorkspace _workspace;
+        private SQLitePersistentStorageService _storageService;
+        private IChecksummedPersistentStorage _storage;
+        private Document _document;
+        private Random _random;
+
+        [GlobalSetup]
+        public void GlobalSetup()
+        {
+            _useExportProviderAttribute.Before(null);
+
+            if (_workspace != null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            _workspace = TestWorkspace.Create(
+@"<Workspace>
+    <Project Language=""NoCompilation"" CommonReferences=""false"">
+        <Document>
+            // a no-compilation document
+        </Document>
+    </Project>
+</Workspace>");
+
+            // Ensure we always use the storage service, no matter what the size of the solution.
+            _workspace.Options = _workspace.Options.WithChangedOption(StorageOptions.SolutionSizeThreshold, -1);
+
+            _storageService = new SQLitePersistentStorageService(
+                _workspace.Services.GetService<IOptionService>(),
+                new LocationService(),
+                new SolutionSizeTracker());
+
+            _storage = _storageService.GetStorageWorker(_workspace.CurrentSolution);
+            if (_storage == NoOpPersistentStorage.Instance)
+            {
+                throw new InvalidOperationException("We didn't properly get the sqlite storage instance.");
+            }
+
+            Console.WriteLine("Storage type: " + _storage.GetType());
+            _document = _workspace.CurrentSolution.Projects.Single().Documents.Single();
+            _random = new Random(0);
+        }
+
+        [GlobalCleanup]
+        public void GlobalCleanup()
+        {
+            if (_workspace == null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            _document = null;
+            _storage.Dispose();
+            _storage = null;
+            _storageService = null;
+            _workspace.Dispose();
+            _workspace = null;
+
+            _useExportProviderAttribute.After(null);
+        }
+
+        private static readonly byte[] s_bytes = new byte[1000];
```

Comment:
❔ Do we have a way to get a distribution of the size of writes we expect to see for e.g. Roslyn?

Reply:
it depends on the index and if we're in teh building phase or not.  indices are rewritten depending on how sensitive they are to checksum changes.  So, for example, our syntactic indices are extremely read-heavy and light on writes (unless, of course, you do something to touch a ton of files on disk).

however, semantic indices are very write heavy as many changes can cause a cascading "recompute teh index" pass to happen when something changes.  

Last time i looked into this there were patterns all over the place.

Reply:
I'm specifically interested in the _size_ of the writes. We use an array pool for write buffers if the writes are less than 4K (with this benchmark fitting inside that pool), but we don't pool arrays for larger writes. I'm curious what size writes we have in practice.